### DISCUSSION TITLE
found python version of mnist code

### DISCUSSION THREAD

**Discussion by Abhinav-Kumar012** [Abhinav-Kumar012/Rust_Python_ML_PE#4] [General]
looks like this the code : [https://github.com/pytorch/examples/tree/main/mnist](https://github.com/pytorch/examples/tree/main/mnist)

Rust version : [https://burn.dev/books/burn/basic-workflow/index.html](https://burn.dev/books/burn/basic-workflow/index.html)

check @valmikGit @Transyltooniaa

**Comment by Transyltooniaa** [comment]
lgtm! will discuss this tomorrow

**Comment by Transyltooniaa** [comment]
@Abhinav-Kumar012  the architecture is different 

| Aspect           | PyTorch               | Burn                     |
|------------------|-----------------------|--------------------------|
| Conv channels    | 1 → 32 → 64           | 1 → 8 → 16               |
| Activation       | After each conv       | After conv2 only         |
| Pooling          | MaxPool2d (2×2)       | AdaptiveAvgPool2d (8×8)  |
| Flatten size     | 9216                  | 1024                     |
| FC hidden size   | 128 (fixed)           | Configurable             |
| Dropout          | Two, different rates  | One, reused              |
| Output           | log_softmax           | logits                   |

**Comment by Abhinav-Kumar012** [comment]
Rust Result:
```
======================== Learner Summary ========================
Model:
Model {
  conv1: Conv2d {ch_in: 1, ch_out: 8, stride: [1, 1], kernel_size: [3, 3], dilation: [1, 1], groups: 1, padding: Valid, params: 80}
  conv2: Conv2d {ch_in: 8, ch_out: 16, stride: [1, 1], kernel_size: [3, 3], dilation: [1, 1], groups: 1, padding: Valid, params: 1168}
  pool: AdaptiveAvgPool2d {output_size: [8, 8]}
  dropout: Dropout {prob: 0.5}
  linear1: Linear {d_input: 1024, d_output: 512, bias: true, params: 524800}
  linear2: Linear {d_input: 512, d_output: 10, bias: true, params: 5130}
  activation: Relu
  params: 531178
}
Total Epochs: 10
```


| Split | Metric                           | Min.     | Epoch    | Max.     | Epoch    |
|-------|----------------------------------|----------|----------|----------|----------|
| Train | Accuracy                         | 83.232   | 1        | 97.603   | 10       |
| Train | CPU Memory                       | 7.333    | 1        | 7.526    | 6        |
| Train | CPU Temperature                  | 82.667   | 10       | 85.402   | 3        |
| Train | CPU Usage                        | 7.361    | 8        | 7.853    | 1        |
| Train | FBetaScore (1) @ TopK(1) [Macro] | 81.526   | 1        | 97.368   | 10       |
| Train | Loss                             | 0.075    | 10       | 0.585    | 1        |
| Train | Precision @ TopK(1) [Macro]      | 83.807   | 1        | 97.600   | 10       |
| Train | Recall @ TopK(1) [Macro]         | 82.980   | 1        | 97.626   | 10       |
| Train | Top-K Accuracy @ TopK(5)         | 97.913   | 1        | 99.965   | 10       |
| Valid | Accuracy                         | 93.720   | 1        | 98.490   | 10       |
| Valid | CPU Memory                       | 7.335    | 1        | 7.530    | 6        |
| Valid | CPU Temperature                  | 84.917   | 9        | 87.783   | 3        |
| Valid | CPU Usage                        | 10.312   | 4        | 11.961   | 6        |
| Valid | FBetaScore (1) @ TopK(1) [Macro] | 93.047   | 1        | 98.372   | 10       |
| Valid | Loss                             | 0.046    | 10       | 0.209    | 1        |
| Valid | Precision @ TopK(1) [Macro]      | 93.840   | 1        | 98.535   | 10       |
| Valid | Recall @ TopK(1) [Macro]         | 93.525   | 1        | 98.525   | 10       |
| Valid | Top-K Accuracy @ TopK(5)         | 99.810   | 1        | 100.000  | 6        |

**Comment by valmikGit** [comment]
<img width="677" height="286" alt="image" src="https://github.com/user-attachments/assets/f2b5fdf3-f9eb-4f42-83aa-08ad6a59ae0d" />

For running the codes.

**Reply by valmikGit** [reply]
This is for running the PyTorch and Tensorflow implementations for the MNIST dataset.

**Comment by valmikGit** [comment]
<img width="997" height="654" alt="image" src="https://github.com/user-attachments/assets/2505fc0a-9980-4ab8-bf72-a9a35b21f204" />
This is for the PyTorch implementation of the text classification task.

**Reply by valmikGit** [reply]
We just have to check whether the entire dataset is being used for training and testing. Currently, I think, the PyTorch implementation is using 2 sample datapoints for training.

**Comment by valmikGit** [comment]
<img width="1006" height="589" alt="image" src="https://github.com/user-attachments/assets/faf7a315-4506-41bb-a928-de0968a49b3b" />

This is for the Tensorflow implementation of the text classification task.

**Reply by valmikGit** [reply]
We just have to check whether the entire dataset is being used for training and testing. Currently, I think, the Tensorflow implementation is using 2 sample datapoints for training.

**Comment by Abhinav-Kumar012** [comment]
@Transyltooniaa @valmikGit please check code before commiting and pushing
you can watch the rust code like nodemon by using 
```sh
bacon run-release -- mnist_ml
```
this command needs to be run at `rust_ml` directory
replace mnist_ml when running other packages

**Reply by Abhinav-Kumar012** [reply]
you can install bacon with 
```sh
cargo install --locked bacon
```
website : [https://dystroy.org/bacon/](https://dystroy.org/bacon/)

**Comment by Abhinav-Kumar012** [comment]
@valmikGit 
this is the command for running mnist
```sh
cargo run -p mnist_ml --release
```

**Comment by Abhinav-Kumar012** [comment]
rust benchmark code written in `python_ml/benchmark` folder
install locust : `pip install locust`
docker build :  `docker build -t mnist -f ./Dockerfile.mnist_inf .`
docker container up : `docker run -p 9050:9050 --name rust_mnist_infer mnist:latest`
locust run : `locust -f ./rust_benchmark.py`

please check @valmikGit @Transyltooniaa

**Reply by Abhinav-Kumar012** [reply]
Please create a folder called `test_data` in `python_ml/benchmark` folder and put some images preferably more that 20
